go/ql/test/library-tests/semmle/go/security/FileSystemAccess/main.go

```diff
@@ -0,0 +1,145 @@
+//go:generate depstubber -vendor  github.com/gofiber/fiber/v2 Ctx New
+//go:generate depstubber -vendor  github.com/kataras/iris/v12/context Context
+//go:generate depstubber -vendor  github.com/labstack/echo/v4 Context New
+//go:generate depstubber -vendor  github.com/spf13/afero Afero,RegexpFs,HttpFs,ReadOnlyFs,MemMapFs,OsFs,BasePathFs WriteReader,SafeWriteReader,WriteFile,ReadFile,ReadDir,NewOsFs,NewRegexpFs,NewReadOnlyFs,NewCacheOnReadFs,,NewHttpFs,NewBasePathFs,NewIOFS
+
+import (
+	"fmt"
+	beego "github.com/beego/beego/v2/server/web"
+	BeegoContext "github.com/beego/beego/v2/server/web/context"
+	"github.com/gin-gonic/gin"
+	"github.com/gofiber/fiber/v2"
+	"github.com/kataras/iris/v12/context"
+	"github.com/labstack/echo/v4"
+	"github.com/spf13/afero"
+	"net/http"
+	"os"
+	"regexp"
+)
+
+func main() {
+	return
+}
+
+func BeegoController(beegoController beego.Controller) {
+	beegoOutput := BeegoContext.BeegoOutput{}
+	beegoOutput.Download("filepath", "license.txt") // $ FileSystemAccess="filepath"
+	buffer := make([]byte, 10)
+	_ = beegoController.SaveToFileWithBuffer("filenameExistsInForm", "filepath", buffer) // $ FileSystemAccess="filepath"
+}
+
+func Afero(writer http.ResponseWriter, request *http.Request) {
+	filepath := request.URL.Query()["filepath"][0]
+	//osFS := afero.NewMemMapFs()
+	// OR
+	osFS := afero.NewOsFs()
+	fmt.Println(osFS.MkdirAll(filepath, 0755))                                   // $ FileSystemAccess=filepath
+	fmt.Println(afero.WriteFile(osFS, filepath, []byte("this is me a !"), 0755)) // $ FileSystemAccess=filepath
+	content, _ := afero.ReadFile(osFS, filepath)                                 // $ FileSystemAccess=filepath
+	fmt.Println(string(content))
+	fmt.Println(osFS.Open(filepath)) // $ FileSystemAccess=filepath
+	// NOT OK
+	fmt.Println(afero.SafeWriteReader(osFS, filepath, os.Stdout)) // $ FileSystemAccess=filepath
+	fmt.Println(afero.WriteReader(osFS, filepath, os.Stdout))     // $ FileSystemAccess=filepath
+
+	// RegexpFs ==> NOT OK
+	fmt.Println("RegexpFs:")
+	regex, _ := regexp.Compile(".*")
+	regexpFs := afero.NewRegexpFs(osFS, regex)
+	fmt.Println(afero.ReadFile(regexpFs, filepath)) // $ FileSystemAccess=filepath
+
+	// ReadOnlyFS ==> NOT OK
+	fmt.Println("ReadOnlyFS:")
+	readOnlyFS := afero.NewReadOnlyFs(osFS)
+	fmt.Println(afero.ReadFile(readOnlyFS, filepath)) // $ FileSystemAccess=filepath
+
+	// CacheOnReadFs ==> NOT OK
+	fmt.Println("CacheOnReadFs:")
+	cacheOnReadFs := afero.NewCacheOnReadFs(osFS, osFS, 10)
+	fmt.Println(afero.ReadFile(cacheOnReadFs, filepath)) // $ FileSystemAccess=filepath
+
+	// HttpFS ==> NOT OK
+	fmt.Println("HttpFS:")
+	httpFs := afero.NewHttpFs(osFS)
+	httpFile, _ := httpFs.Open(filepath) // $ FileSystemAccess=filepath
+	tmpbytes := make([]byte, 30)
+	fmt.Println(httpFile.Read(tmpbytes))
+	fmt.Println(string(tmpbytes))
+
+	// osFS ==> NOT OK
+	fmt.Println("Afero:")
+	afs := &afero.Afero{Fs: osFS}       // $ succ=&... pred=osFS
```

Comment:
Please check this pattern also works.
```suggestion
	afs0 := afero.Afero{Fs: osFS}
	afs := &afs0
```